FTP Dynamic Connection Example (#1545)
mulesoft/docs-connectors: ftp/1.5/modules/ROOT/pages/ftp-dynamic-connection.adoc

> = Connect to FTP Servers Dynamically Example - Mule 4

Reviewer: ```suggestion
= Connect Dynamically to FTP Servers Example - Mule 4
```

Author: This is based on SFTP connectors. We are keeping a parallel with it.

> Many integrations require connecting to different servers depending on a certain condition.

Reviewer: ```suggestion
Integrations sometimes require connecting to different servers under different conditions.
```

Examples of this include:

> * Connecting to different invoice storage servers depending on the branch that emits an invoice.

Reviewer: ```suggestion
* Connecting to different invoice storage servers depending on the branch that sends an invoice.
```

* Connecting to different servers depending on an integration subject, such as in a multitenant use case.

> To accommodate these use cases, the `config` element supports expressions, which allows for connection parameters to evaluate these conditions and connect to the correct server.

Reviewer: ```suggestion
To accommodate these conditions, the `config` element supports expressions, which allow for connection parameters to evaluate these conditions and connect to the correct server.
```

Author: I'll use the same verbiage that was used in the SFTP connectors.

The following example shows a dynamic multicast application that:

> * Defines a config for the connector in which the `host`, `username`, and `password` are expressions.

Reviewer: ```suggestion
* Defines a configuration for the connector in which the `host`, `username`, and `password` are expressions.
```

Reviewer: I would not use bold for any of the names in this list, because in this context they seem to refer to the general names of these things rather than actual UI labels. Please doublecheck with Kevin, because Connectors docs does this a bit differently than other docs.

* Describes a flow, triggered by an *HTTP Listener*, in which content is posted through HTTP.

> * Uses the File Connector *Read* operation to load a recipients CSV file that contains a set of FTP destinations, with columns such as `host`, `user`, and `port`.

Reviewer: ```suggestion
* Uses the File Connector *Read* operation to load a recipient's CSV file that contains a set of FTP destinations, with columns such as `host`, `user`, and `port`.
```

Author: "recipients" is the name of the file. I'll change it accordingly.

* Uses a *For Each* component and an FTP *Write* operation to write the contents into each of the FTP destinations.

> ** On each *For Each* iteration, the expressions set in the FTP global configuration element resolve to a different value and establish different connections to each of the servers.

Reviewer: ```suggestion
* Uses a *For Each* component and an FTP *Write* operation to write the contents into each of the FTP destinations. On each *For Each* iteration, the expressions set in the FTP global configuration element resolve to a different value and establish different connections to each of the servers.
```

> To test this example, create the Mule application, and run and test the application with curl commands:

Reviewer: I would omit the second use of "test" because that is the goal of all of these actions taken together:
```suggestion
To test this example, create the Mule application and run it using curl commands:
```

image::ftp_dynamic_connect_flow.png[FTP Servers Dynamic Connection Flow in Studio Canvas]

== Create the Mule Application

> To create the Mule flow:

Reviewer: application or flow?
Avoid using more than one term to mean one thing. GLOBALLY in all writing.

Author: I'll change to "Mule application flow"

. In the *Mule Palette* view, select *HTTP > Listener*.
. Drag *HTTP Listener* to the Studio canvas.
. On the *HTTP Listener* configuration screen, optionally change the value of the *Display Name* field.
. Set *Path* to `/multitenant`.
. Click the plus sign (*+*) next to the *Connector configuration* field to configure a global element that can be used by all instances of HTTP Listener in the app.
. On the *General* tab, set the following fields:

> +

Reviewer: I don't think you need to format this space?

* *Host*: `All Interfaces [0.0.0.0] (default)`
* *Port*: `8081`

> +

Reviewer: ditto comment Line 34?

. Click *OK*.
. Drag a *Set Variable* component to the right of the *HTTP Listener* source.
. Set *Name* to `content` and *Value* to `#[payload]`.
. Drag the File Connector *Read* operation to the right of the *Set Variable* component.
. Set the *File Path* field to `recipients.csv`. +

>   This reads a CSV file that contains a random set of FTP destinations with columns such as `host`, `user`, and `port`.

Reviewer: ```suggestion
. Set the *File Path* field to `recipients.csv`. This reads a CSV file that contains a random set of FTP destinations with columns such as `host`, `user`, and `port`.
```

. Click the plus sign (*+*) next to the *Connector configuration* field to configure a global element that can be used by all instances of File Connector in the app.
. Accept the default settings and click *OK*.
. On the *MIME Type* tab of the *Read* operation, for *MIME Type*, select *application/csv*.

> 

Reviewer: omit space

. In Studio, drag a *Transform Message* component to the right of the *Read* operation.
. In the *Output* section of the component, add the following DataWeave code to map the columns of the CSV file:
+
.DataWeave script:
[source,dataweave,linenums]
----
%dw 2.0
output application/java
---
payload map using (row = $) {
   host: row.Host,
   user: row.User,
   password: row.Password}
----
+
. Drag a *For Each* component to the right of the *Transform Message* component.

> . Drag an FTP *Write* operation into the *For Each* component. +
> On each For Each iteration, the *Write* operation writes the contents into each of the FTP destinations.

Reviewer: ```suggestion
. Drag an FTP *Write* operation into the *For Each* component. On each For Each iteration, the *Write* operation writes the contents into each of the FTP destinations.
```

. Click the plus sign (*+*) next to the *Connector configuration* field to configure a global element that can be used by all instances of FTP *Write* operation in the app, and set the following fields:

> +

Reviewer: omit

* *Host*: `#[payload.host]`
* *Username*: `#[payload.username]`
* *Password*: `#[payload.password]`

> +

Reviewer: omit

[start=6]
. Click *OK*.
. Set *Path* to `demo.txt`.
. Set the *Content* field to `payload`.
. Drag a *Set Payload* component to the right of the *For Each* component.
. Set the *Value* field to `Multicast Ok`.

> == Run and Test your Mule Application

Reviewer: ```suggestion
== Run and Test Your Mule Application
```

Reviewer: Are we running it with the purpose of testing it?
Otherwise, it is more logical to test first and then run...or deploy.

After creating your Mule application, run and test it:

> . In Studio, save your Mule app.

Reviewer: Shouldn't saving the app be the final step of creating it?

. Click the project name in *Package Explorer* and then click *Run* > *Run As* > *Mule Application*.
. Open a browser and type `+http://0.0.0.0:8081/multitenant+`. +

> Note about Mule 4 behavior for Mule 3 users:

Reviewer: Is the following text for Mule 3 users an addendum to the task that is the focus of this topic?
Does the task end at Step 2, or are these listed items in relation to Step 2 itself?
And what does this intro fragment mean: if one is a Mule 3 user, why does one need to know about Mule 4 behavior...or do you mean to point out to users who have upgraded to Mule 4?
Maybe what you really mean is...:
```suggestion
If you are accustomed to running and testing in Mule 3, note the following changes in Mule 4:
```

> * The information posted through the HTTP Listener source is written to each FTP site multiple times.
> +
> Because the listener makes use of the xref:mule-runtime::streaming-about.adoc[repeatable streams feature], you do not need to worry about consuming the stream multiple times.

Reviewer: ```suggestion
* Although the information posted through the HTTP Listener source is written to each FTP site multiple times, the repeatable streams feature in Mule 4 prevents the overhead of consuming the stream multiple times.
```

> * The For Each component automatically goes through each line of the CSV file.

Reviewer: ```suggestion
* In Mule 4, the For Each component automatically goes through each line of the CSV file without needing to first transform it into a Java structure.
```

+

> In Mule 3, you first needed to transform the CSV file into a Java structure, but because Mule 4 is now Java agnostic, this works out-of-the-box.

Reviewer: See suggested rewording, Line 93.

> When reading or listing files, you might be interested in the file's metadata (for example, the file name, full path, size, timestamp). The connector uses the Mule message attributes to access this information.

Reviewer: How does this relate to the focus on this topic, running and testing?

* Connecting to different servers depending on an integration subject, such as in a multitenant use case.

To accommodate these use cases, the `config` element supports expressions, which allows for connection parameters to evaluate these conditions and connect to the correct server.

The following example shows a dynamic multicast application that:

* Defines a config for the connector in which the `host`, `username`, and `password` are expressions.
* Describes a flow, triggered by an *HTTP Listener*, in which content is posted through HTTP.
* Uses the File Connector *Read* operation to load a recipients CSV file that contains a set of FTP destinations, with columns such as `host`, `user`, and `port`.
* Uses a *For Each* component and an FTP *Write* operation to write the contents into each of the FTP destinations.
** On each *For Each* iteration, the expressions set in the FTP global configuration element resolve to a different value and establish different connections to each of the servers.

To test this example, create the Mule application, and run and test the application with curl commands:

> image:ftp_dynamic_connect_flow.png:[FTP Servers Dynamic Connection Flow in Studio Canvas]

Reviewer 2: ```suggestion
image::ftp_dynamic_connect_flow.png[FTP Servers Dynamic Connection Flow in Studio Canvas]
```

== Create the Mule Application

To create the Mule flow:

. In the *Mule Palette* view, select *HTTP > Listener*.
. Drag *HTTP Listener* to the Studio canvas.
. On the *HTTP Listener* configuration screen, optionally change the value of the *Display Name* field.

> . Set the *Path* field to `/multitenant`

Reviewer 2: ```suggestion
. Set *Path* to `/multitenant`.
```

Reviewer 2: Per Style guide now consider referring to UI elements by label rather than element type:

Before
. Select the Upload all my files now checkbox and click the Next button.

After
. Select Upload all my files now and click Next.

. Click the plus sign (*+*) next to the *Connector configuration* field to configure a global element that can be used by all instances of HTTP Listener in the app.
. On the *General* tab, set the following fields:
+
* *Host*: `All Interfaces [0.0.0.0] (default)`
* *Port*: `8081`
+
. Click *OK*.
. Drag a *Set Variable* component to the right of the *HTTP Listener* source.

> . Set the *Name* field to `content` and the *Value* field to `#[payload]`.

Reviewer 2: ```suggestion
. Set *Name* to `content` and *Value* to `#[payload]`.
```

. Drag the File Connector *Read* operation to the right of the *Set Variable* component.

> . Set the *File Path* field to `recipients.csv`. +

Reviewer 2: DITTO

  This reads a CSV file that contains a random set of FTP destinations with columns such as `host`, `user`, and `port`.
. Click the plus sign (*+*) next to the *Connector configuration* field to configure a global element that can be used by all instances of File Connector in the app.
. Accept the default settings and click *OK*.

> . On the *MIME Type* tab of the *Read* operation, for the *MIME Type* field, select `application/csv`.

Reviewer 2: ```suggestion
. On the *MIME Type* tab of the *Read* operation, for *MIME Type*, select *application/csv*.
```

> === Configure Additional Components and Operations
> 
> Continue creating your Mule application by adding additional components and operations:
> 

Reviewer 2: ```suggestion

```

Reviewer 2: Remove this division, you can join all steps together.

. In Studio, drag a *Transform Message* component to the right of the *Read* operation.
. In the *Output* section of the component, add the following DataWeave code to map the columns of the CSV file:
+
.DataWeave script:
[source,dataweave,linenums]
----
%dw 2.0
output application/java
---
payload map using (row = $) {
   host: row.Host,
   user: row.User,
   password: row.Password}
----
+
. Drag a *For Each* component to the right of the *Transform Message* component.
. Drag an FTP *Write* operation into the *For Each* component. +
On each For Each iteration, the *Write* operation writes the contents into each of the FTP destinations.

> . Click the plus sign (*+*) next to the *Connector configuration* field to configure a global element that can be used by all instances of FTP Write operation in the app, and set the following fields:

Reviewer 2: ```suggestion
. Click the plus sign (*+*) next to the *Connector configuration* field to configure a global element that can be used by all instances of FTP *Write* operation in the app, and set the following fields:
```

Reviewer 2: Operation, field, and tab names are styled in bold

+
* *Host*: `#[payload.host]`
* *Username*: `#[payload.username]`
* *Password*: `#[payload.password]`
+
[start=6]
. Click *OK*.

> . On the configuration screen, set the *Path* field to `demo.txt`.

Reviewer 2: ```suggestion
. Set *Path* to `demo.txt`.
```

> . Set the *Content* field to `payload`.

Reviewer 2: DITTO

. Drag a *Set Payload* component to the right of the *For Each* component.

> . Set the *Value* field to `Multicast Ok`.

Reviewer 2: DITTO
